File: src/state/pair.js
'use strict';

/**
 * Monta a mensagem auxiliar de 12 bytes (indica direção e flags)
 * F0 43 10 3E 7F 11 [RESET_FLAG] 00 [SOURCE] 00 [TARGET] F7
 */
function buildAuxMsg(resetFlag, sourceCh, targetCh) {
    return [0xF0, 0x43, 0x10, 0x3E, 0x7F, 0x11,
            resetFlag, 0x00, sourceCh, 0x00, targetCh, 0xF7];
}

/**
 * Monta a mensagem de estado de 14 bytes (liga/desliga o pair)
 * F0 43 10 3E 7F 01 18 00 [CH_BYTE] 00 00 00 [STATE] F7
 * CH_BYTE = sempre o menor índice do par (ex: CH1=0, CH3=2)
 */
function buildStateMsg(chByte, state) {
    return [0xF0, 0x43, 0x10, 0x3E, 0x7F, 0x01, 0x18,
            0x00, chByte, 0x00, 0x00, 0x00, state, 0xF7];
}

/** Retorna o CH_BYTE = menor índice (canal ímpar da UI) */
function getChByte(chA, chB) {
    return Math.min(chA, chB);
}

/**
 * Ativa pair entre chA e chB, copiando dados de sourceCh para o outro.
 * Envia: mensagem auxiliar (12b) + mensagem de estado ON (14b)
 */
function pairChannels(midiOutput, chA, chB, sourceCh) {
    const targetCh = (sourceCh === chA) ? chB : chA;
    const chByte = getChByte(chA, chB);
    
    // 1. Enviar mensagem auxiliar com a direção da cópia
    midiOutput.sendMessage(buildAuxMsg(0x00, sourceCh, targetCh));
    
    // 2. Enviar mensagem de ativação do Pair
    midiOutput.sendMessage(buildStateMsg(chByte, 0x01));
    
    console.log(`[MIDI] Pair Ativado: CH ${chA+1} + CH ${chB+1} (Source: CH ${sourceCh+1})`);
}

/**
 * Desativa pair entre chA e chB.
 * Envia apenas: mensagem de estado OFF (14b)
 */
function unpairChannels(midiOutput, chA, chB) {
    const chByte = getChByte(chA, chB);
    midiOutput.sendMessage(buildStateMsg(chByte, 0x00));
    console.log(`[MIDI] Pair Desativado: CH ${chA+1} + CH ${chB+1}`);
}

/**
 * Reseta ambos os canais e mantém o pair ativo.
 * Envia: mensagem auxiliar com RESET_FLAG=1 (12b) + mensagem de estado ON (14b)
 */
function resetBothChannels(midiOutput, chA, chB) {
    const chByte = getChByte(chA, chB);
    midiOutput.sendMessage(buildAuxMsg(0x01, chA, chB));
    midiOutput.sendMessage(buildStateMsg(chByte, 0x01));
    console.log(`[MIDI] Reset Both: CH ${chA+1} e CH ${chB+1}`);
}

module.exports = { pairChannels, unpairChannels, resetBothChannels };

File: server_rust/src/midi/pair.rs
pub fn build_aux_msg(reset_flag: u8, source_ch: u8, target_ch: u8) -> Vec<u8> {
    vec![
        0xF0, 0x43, 0x10, 0x3E, 0x7F, 0x11, reset_flag, 0x00, source_ch, 0x00, target_ch, 0xF7,
    ]
}

pub fn build_state_msg(ch_byte: u8, state: u8) -> Vec<u8> {
    vec![
        0xF0, 0x43, 0x10, 0x3E, 0x7F, 0x01, 0x18, 0x00, ch_byte, 0x00, 0x00, 0x00, state, 0xF7,
    ]
}

fn get_ch_byte(ch_a: u8, ch_b: u8) -> u8 {
    ch_a.min(ch_b)
}

pub fn build_pair(ch_a: u8, ch_b: u8, source_ch: u8) -> (Vec<u8>, Vec<u8>) {
    let target_ch = if source_ch == ch_a { ch_b } else { ch_a };
    let ch_byte = get_ch_byte(ch_a, ch_b);
    (
        build_aux_msg(0x00, source_ch, target_ch),
        build_state_msg(ch_byte, 0x01),
    )
}

pub fn build_unpair(ch_a: u8, ch_b: u8) -> Vec<u8> {
    let ch_byte = get_ch_byte(ch_a, ch_b);
    build_state_msg(ch_byte, 0x00)
}

pub fn build_reset(ch_a: u8, ch_b: u8) -> (Vec<u8>, Vec<u8>) {
    let ch_byte = get_ch_byte(ch_a, ch_b);
    (
        build_aux_msg(0x01, ch_a, ch_b),
        build_state_msg(ch_byte, 0x01),
    )
}
